### ISSUE TITLE
Can the server be compiled to JavaScript for use in Deno?

### ISSUE THREAD

**Issue by guest271314** [BiagioFesta/wtransport#40]

**Comment by MOZGIII** [comment]
You can use this crate to build a deno plugin (see https://deno.land/manual@v1.35.3/runtime/ffi_api).
It is also possible to build a node.js native extension using this crate!

**Comment by guest271314** [comment]
I'll have to check that out. QuickJS provides a similar means to import a shared object file into a context. I have not had much success building using Rust on a temporary file system. I'm interested in making this so using JavaScript alone. I'll give the Deno FFI interace a try anyway. Thanks.

**Comment by guest271314** [comment]
@MOZGIII Where do I start?

**Comment by MOZGIII** [comment]
Probably, create a new repo 😄
Then, I suggest checking out this: https://blog.logrocket.com/how-to-create-a-deno-plugin-in-rust/ and https://alexandrempsantos.com/deno/creating-a-deno-plugin/ (googled it in 1 minute) - there seems to be a special path for deno plugins specifically.

Next up, you'll have to wrap the interface of the `wrtransport` with your glue code to expose each relevant object and fn to `deno`.

If you would like to go through the more generic `ffi` route - then start by exposing `wtransport` calls/types in C via `cbindgen` or what you prefer, then wrap them into the `ffi` interface from the deno end. This way you can create binding for a lot of platforms that can do ffi: nodejs, ruby, java, python, etc... It is more difficult though, as you'll need to sort out the async calls interface - deno plugins have special support for that.

**Comment by guest271314** [comment]
Assuming I have a total of 1.1 GB of space on a temporary file system running on RAM only won't installing Rustuse all of that space before we get to importing crates?

**Comment by MOZGIII** [comment]
Oof, yeah, that would not be enough space most likely.

Actually, my local checkout of `wtransport` is ~650 mb. Might be feasible.

**Comment by guest271314** [comment]
Doesn't Rust need to be installed before we proceed to check out the `wtransport` crate? That is what has been somewhat challenging to use Rust for anything. If I install `build-essential` or `clang` then I'm pretty much locked in to C and C++ due to the environment I'm using. The last time I checked the recommended installation of Rust is over 1GB alone.

**Comment by MOZGIII** [comment]
Oh, yes, in that sense you'll be stuck.

I'm curious about your setup now, why do you have such restrictions?

**Comment by guest271314** [comment]
I usually run a Linux live ISO on a 4-16 GB medium. The idea is a portable, embeddable system.

**Comment by guest271314** [comment]
What I am basically trying to achieve is this https://bugs.chromium.org/p/chromium/issues/detail?id=1214621. I don't really need networking, however that is where WHATWG Streams Standard is implemented.

Using a `WebSocketStream` works to an appreciable degree, when using `ws:`. Deno's `Deno.listenTls()` is broken, times out in ~5 minutes, so that is a concern if I'm trying to use `wss:`.

**Comment by MOZGIII** [comment]
Looks like you might as well just write a full-blown application to expose stdio streams with WebTransport in rust - why bother with deno? Maybe even use ready-made `socat` or something similar if it does the job.

**Comment by guest271314** [comment]
> why bother with deno?

Good question.

So far Deno is the JavaScript runtime I have found implements full-duplex streaming using `fetch()` https://github.com/guest271314/native-messaging-deno/tree/fetch-duplex.

I've created various workarounds to stream to and from the browser locally. 

From my vantage one has to commit to using Rust. Which is fine if I have a roadmap to implement when I freeze the OS doing some experiments and start over, which happens every few days. How would you go about implementing the details mentioned in https://bugs.chromium.org/p/chromium/issues/detail?id=1214621, and https://bugs.chromium.org/p/chromium/issues/detail?id=248548 using Rust - without Chromium complicity?

Seems like somebody over in Deno world just banned me from their Discord channel, without notice. Perhaps because I mentioned `Deno.listenTls()` is broken and nobody is fixing it? I guess if the policy or practice is to ban people who call out outstanding bugs wishing them to the corn field wishes away the bugs? Too funny.

**Comment by guest271314** [comment]
I guess using JSON over IPC ain't that bad. I've managed to just use the valid JSON `[0, 255, ...]` to send to and from the Native Messaging host. 

I have been persistently interested in implementing a WHATWG Streams Standard version of this https://github.com/guest271314/captureSystemAudio/tree/master/native_messaging/capture_system_audio - primarily to compare streams to JSON. 

I'll close this and won't hold up anybody's time over here. Thanks for the suggestions. Cheers.

**Comment by MOZGIII** [comment]
You could try shared memory instead of streaming... Or at the very least use bytearrays instead of JSON. I'd estimate the JSON encoding/decoding to cost around 5-30 ms nowadays for something like this, for audio it might be on the edge of what's good.
Btw, I was working on https://github.com/MOZGIII/netsound as a building block for something like this.

**Comment by guest271314** [comment]
> You could try shared memory instead of streaming...

I thought about that. Using `WebAssembly.Memory` and `grow()` https://github.com/guest271314/quictransport/blob/main/audioworklet-webassembly-memory-grow/quicTransportAudioWorkletMemoryGrow.js. I don't think there's any way to share a `WebAssembly.Memory` object between a non-browser context and a browser context https://github.com/wasmerio/wasmer-php/issues/121.

JSON works for both capturing and streaming audio, voice, and speech synthesis playback. I've captured and recorded over 1 hour of live audio output to speakers or headphones, piped to a `MediaStreamTrackGenerator` for the ability to share the media stream with peers and simultaneously encoded the audio to an MP3 in the browser.

I am just trying to test/compare JSON implementation with streams implementation, to derive empirical results.

**Comment by guest271314** [comment]
> I'd estimate the JSON encoding/decoding to cost around 5-30 ms nowadays for something like this

Technically we don't really have to encode or decode JSON sent to the browser from a Native Messaging host. The data in the callback has the literal form `[0, 255, ...]`, in JavaScript a JavaScript plain object, not a string `"[0, 255, ...]"`. To pass that value to the `writable` side of a `TransformStream` we do have to pass that array to a `Uint8Array` constructor, something like https://github.com/guest271314/captureSystemAudio/blob/master/native_messaging/capture_system_audio/background.js#L110-L132

```
      async nativeMessageStream() {
        return new Promise((resolve) => {
          this.port = chrome.runtime.connect(this.src.hostname);
          const { readable, writable } = new TransformStream();
          this.stdout = readable;
          this.writer = writable.getWriter();
          this.port.onMessage.addListener(async (value) => {
            try {
              await this.writer.ready;
              await this.writer.write(new Uint8Array(value));
            } catch (e) {
              console.log(e.message);
            } finally {
              return true;
            }
          });
          this.port.onDisconnect.addListener(this.portDisconnectedResolve);
          this.port.postMessage(this.stdin);
          resolve(this.captureSystemAudio());
        }).catch((err) => {
          throw err;
        });
      }
```

**Comment by guest271314** [comment]
> As in for wasi, etc? In theory yes, but there might be technical difficulties.

Since then I've tried to compile fail-components/webtransport to WASM. Stuck on protobuffer and subprocesses. Somebody put this https://github.com/maceip/quinn-wasm/tree/direct-sockets together, though it's WIP, doesn't work. Right now I'm trying to figure out how to send data back to WebTransport client from this https://github.com/colocohen/quico, which is all JavaScript. 

I've got `UDPSocket` in the browser. I just need to make to TLS connection. And implement the HTTP/3 parts in JavaScript or compile to WASM from source code that was designed with that capability in mind.
